docs/fsharp/style-guide/component-design-guidelines.md

```diff
@@ -0,0 +1,884 @@
+
+#### Use either PascalCase or camelCase for public functions and values in F# modules
+
+camelCase is used for public functions that are designed to be used unqualified (for example, `invalidArg`), and for the “standard collection functions” (for example, List.map). In both these cases, the function names act much like keywords in the language.
+
+### Object, Type, and Module design
+
+#### Use namespaces or modules to contain your types and modules
+
+Each F# file in a component should begin with either a namespace declaration or a module declaration.
+
+```fsharp
+namespace Fabrikam.BasicOperationsAndTypes
+
+type ObjectType1() =
+    ...
+
+type ObjectType2() =
+     ...
+
+module CommonOperations =
+    ...
+```
+
+or
+
+```fsharp
+module Fabrikam.BasicOperationsAndTypes
+
+type ObjectType1() =
+    ...
+
+type ObjectType2() =
+    ...
+
+module CommonOperations =
+    ...
+```
+
+The differences between using modules and namespaces to organize code at the top level are as follows:
+
+* Namespaces can span multiple files
+* Namespaces cannot contain F# functions unless they are within an inner module
+* The code for any given module must be contained within a single file
+* Top-level modules can contain F# functions without the need for an inner module
+
+The choice between a top-level namespace or module affects the compiled form of the code, and thus will affect the view from other .NET languages should your API eventually be consumed outside of F# code.
+
+#### Use methods and properties for operations intrinsic to object types
+
+When working with objects, it is best to ensure that consumable functionality is implemented as methods and properties on that type.
+
+```fsharp
+type HardwareDevice() =
+
+    member this.ID = ...
+
+    member this.SupportedProtocols = ...
+
+type HashTable<'Key,'Value>(comparer: IEqualityComparer<'Key>) =
+
+    member this.Add(key, value) = ...
+
+    member this.ContainsKey(key) = ...
+
+    member this.ContainsValue(value) = ...
+```
+
+The bulk of functionality for a given member need not necessarily be implemented in that member, but the consumable piece of that functionality should be.
+
+#### Use classes to encapsulate mutable state
+
+In F#, this only needs to be done where that state is not already encapsulated by another language construct, such as a closure, sequence expression, or asynchronous computation.
+
+```fsharp
+type Counter() =
+    // let-bound values are private in classes.
+    let mutable count = 0
+
+    member this.Next() =
+        count <- count + 1
+        count
+```
+
+#### Use interfaces to group related operations
+
+We recommend you use interface types to represent a set of operations. This is preferred to other options, such as tuples of functions or records of functions.
```

Comment:
Avoid "we", unless this is official Microsoft guidance. Consider just "Use interface types...."